utils/lib_govm.sh

```diff
@@ -0,0 +1,224 @@
+    esac
+    echo "${OS}"
+}
+
+go.arch() {
+    local ARCH
+    case "$(command uname -m)" in
+        "x86_64") ARCH=amd64 ;;
+        "aarch64") ARCH=arm64 ;;
+        "armv6" | "armv7l") ARCH=armv6l ;;
+        "armv8") ARCH=arm64 ;;
+        .*386.*) ARCH=386 ;;
+        ppc64*) ARCH=ppc64le ;;
+        *) die 42 "ARCH is unknown: $(command uname -m)" ;;
+    esac
+    echo "${ARCH}"
+}
+
+# Go version management (go.vm)
+# -----------------------------
+
+go.vm.ensure() {
+    if ! go.vm.is_installed; then
+        # shellcheck disable=SC2119
+        go.vm.install
+    fi
+}
+
+go.vm.is_installed() {
+    # is true if "go" command is installed
+    [[ -f "${GOROOT}/bin/go" ]]
+}
+
+# shellcheck disable=SC2120
+go.vm.install() {
+
+    # Go versions manager; to install Go at arbitrary place:
+    #
+    # usage:  go.vm.install <version> <dest>
+
+    local version dest fname sha size tmp
+    version="${1:-$GOVERSION}"
+    dest="${2:-$GOROOT}"
+
+    info_msg "Install Go in ${dest}"
+
+    # HINT: the python requirements needed by go.vm.version are taken from the
+    # developer environment. If it is not yet installed, install it now ..
+    pyenv.install
+
+    # fetch go version ..
+    local buf=()
+    mapfile -t buf < <(
+        go.vm.version "${version}" archive "$(go.os)" "$(go.arch)" filename sha256 size
+    )
+    if [ ${#buf[@]} -eq 0 ]; then
+        die 42 "can't find info of golang version: ${version}"
+    fi
+    fname="${buf[0]}"
+    sha="${buf[1]}"
+    size="$(numfmt --to=iec "${buf[2]}")"
+
+    info_msg "Download go binary ${fname} (${size}B)"
+    cache_download "${_GO_DL_URL}/${fname}" "${fname}"
+
+    pushd "${CACHE}" &>/dev/null
+    echo "${sha}  ${fname}" >"${fname}.sha256"
+    if ! sha256sum -c "${fname}.sha256" >/dev/null; then
+        die 42 "downloaded file ${fname} checksum does not match"
+    else
+        info_msg "${fname} checksum OK"
+    fi
+    popd &>/dev/null
+
+    info_msg "install golang"
+
+    tmp="$(mktemp -d)"
+    tar -C "${tmp}" -xzf "${CACHE}/${fname}"
+    rm -rf "${dest}"
+    mkdir -p "$(dirname "${dest}")"
+    mv "${tmp}/go" "${dest}"
+
+    mkdir -p "$(dirname "$GOENV")"
+    export GOENV
+
+    "${GOVM_EXE}" telemetry off
+    "${GOVM_EXE}" env -w \
+        GOBIN="$GOBIN" \
+        GOTOOLCHAIN="$GOTOOLCHAIN" \
+        GOCACHE="$GOCACHE" \
+        GOPATH="$GOPATH" \
+        GOMODCACHE="$GOMODCACHE"
+
+    mkdir -p "${GOMODCACHE}"
+}
+
+go.vm.list() {
+
+    # Go versions manager; list Go versions (stable)
+
+    "${PY_ENV_BIN}/python" <<EOF
+import sys, json, requests
+resp = requests.get("${_GO_DL_URL}/?mode=json&include=all")
+for ver in json.loads(resp.text):
+    if not ver['stable']:
+        continue
+    for f in ver['files']:
+        if f['kind'] != 'archive' or not f['size'] or not f['sha256'] or len(f['os']) < 2:
+            continue
+        print(" %(version)-10s|%(os)-8s|%(arch)-8s|%(filename)-30s|%(size)-10s|%(sha256)s" % f)
```

Comment:
It'd probably be cleaner to either
- use `curl` and `jq` (or a similar JSON parser)
- put the python scripts in a separate file to benefit from code formatting and other LSP hints

Reply:
> * use `curl` and `jq` (or a similar JSON parser)

The code from above is executed in the pythen dev-env we already have (`"${PY_ENV_BIN}/python"`) .. while `jq` is a dependency that needs to be installed first from distro's package manager ..



> * put the python scripts in a separate file to benefit from code formatting and other LSP hints

:thinking: thats true  .. but involves more files  (laying arround).. I think we will never touch this code again, and its easier to have all the "go" code in one file.  But if you insist, I can separate the python code into its own file.